### ISSUE TITLE
t.w.p.msn does not properly execute errbacks when an error packet is received

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2019]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)| @habnabit reported|
|-|-|
|Trac ID|trac#2019|
|Type|defect|
|Created|2006-08-24 20:43:51Z|

In msn.MSNEventBase, the dict that stores the deferreds uses ints as keys, though when an error packet is received, the dict's has_key is called with a str, not an int. This causes errbacks to not fire.

The patch included fixes the defect.

Attachments:

* [msnpatch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0c4/0c422ba64421103f8f58fc3c8676caf9c7c73178/2a71f96d91bd459fae3704bda97224ed50f955df.diff) (256 bytes) - added by habnabit on 2006-08-24 20:45:10Z - 
* [msnpatch2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0c4/0c422ba64421103f8f58fc3c8676caf9c7c73178/ffb4c912268f448f1faf69ca42295149920c5659.diff) (1026 bytes) - added by habnabit on 2006-08-24 22:20:00Z - Context patch.
* [testpatch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0c4/0c422ba64421103f8f58fc3c8676caf9c7c73178/354671bd3596f4a084f9eef2b7cb781b487923c4.diff) (855 bytes) - added by habnabit on 2006-08-24 22:47:49Z - A patch to test_msn.py that will add a test that fails without the patch to MSNEventBase.
* [msnpatch3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0c4/0c422ba64421103f8f58fc3c8676caf9c7c73178/bd6cdb801ac94b7e28912ec3f9c530ecc41346a2.diff) (1066 bytes) - added by habnabit on 2006-08-25 09:55:54Z - Patch that also calls the errback with an MSNProtocolError instance.
* [testpatch2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0c4/0c422ba64421103f8f58fc3c8676caf9c7c73178/923918f940d9d341072f00ec3f7a67ede7e03179.diff) (912 bytes) - added by habnabit on 2006-08-25 09:56:32Z - Test patch that uses assertEqual.

<details><summary>Searchable metadata</summary>

```
trac-id__2019 2019
type__defect defect
reporter__habnabit habnabit
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1156452231000000 1156452231000000
changetime__1156663292000000 1156663292000000
version__None None
owner__spiv spiv
cc__habnabit cc__mwh cc__spiv
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_1"></a>|@mwhudson commented|
|-|-|

The patch is a straight diff, please supply a unified or a context diff.

Also, a test case would surely be nice :)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_2"></a>|@habnabit commented|
|-|-|

The test is verified to fail under TwistedWords-0.4.0.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|spiv set owner to @habnabit|
|-|-|

You should use `self.assertEqual(a, b, msg)` rather than `self.failUnless(a == b, msg)`.  Also, "testErrback" doesn't really explain the intent of the test.  Add a docstring like "Ensure that errors trigger an errback."

But otherwise the fix is small and makes sense, and the test is also small and tests a fairly minimal unit of code, which is good.  It *might* be a good idea to also assert something about the failure passed to the errback.

I'd be happy to merge this with those changes.  Feel free to reassign to me when you've made the changes.

Thanks for patch, and for providing a test!

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_5"></a>|@habnabit set owner to spiv|
|-|-|

spiv, will you close the ticket when the changes are merged?

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|spiv removed owner|
|-|-|

I realised the test could be made even simpler with assertFailure -- as well as checking that the error received is indeed an MSNProtocolError, as I requested in my first review.  In [branches/msn-errback-2019](branches/msn-errback-2019) I've committed your patches and then a couple of tweaks (simplify test, fitting code in 80-columns as per our style guide, pass a single string to MSNProtocolError rather than a tuple to be consistent with other code) on top of your changes.

Seeing as some of this code is now mine, and I'm not particularly familiar with twisted.words, I'm putting this up for review by another developer before I commit it.  My main concern is I don't really know if MSNProtocolError is the right thing to raise in that case, although it's clearly better than trying to raise an integer :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to spiv|
|-|-|

I think your concern was well-founded.  I would suggest defining a new exception class and passing an instance of it to the errback, instead of using MSNProtocolError, the main purpose of which is to signal a parser bug, rather than an error response to a command (although I have no idea what error responses exist or indicate in this protocol ;).  Since this new exception class won't just be logged and otherwise ignored, but potentially used by application code, I would also suggest breaking with the prevaling convention in this module and constructing it with structured data, rather than a string (ie, just drop the format string).

Other minor points.  As long as you are touching this code, `has_key' may as well become `in'.  And the docstring on testErrback should look like

```
        """
        Ensures that errors trigger an errback.
        """
```

although it should also be more specific and talk about error *responses* and Deferreds returned by commands.

Other than these minor issues, the branch looks good.  The test is indeed quite unit-y and the broken behavior appears to be correct.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|spiv set status to `closed`|
|-|-|

(In [17982]) Merge msn-errback-2019.

* Author: habnabit, spiv
* Reviewer: spiv, exarkun

Fixes [#2019](https://github.com/twisted/twisted/issues/2019)

Handle error responses from the server in reply to commands by firing the
command's deferred with MSNCommandFailed, rather than breaking.

Based on a fix and test submitted by habnabit.
